sql/analyzer/validation_rules.go

```diff
@@ -210,6 +220,40 @@ func validateSchema(t *plan.ResolvedTable) error {
 	return nil
 }
 
+func validateUnionSchemasMatch(ctx *sql.Context, a *Analyzer, n sql.Node) (sql.Node, error) {
+	span, _ := ctx.Span("validate_union_schemas_match")
+	defer span.Finish()
+
+	var firstmismatch []string
+	plan.Inspect(n, func(n sql.Node) bool {
+		if u, ok := n.(*plan.Union); ok {
+			ls := u.Left.Schema()
+			rs := u.Right.Schema()
+			if len(ls) != len(rs) {
+				firstmismatch = []string{
+					fmt.Sprintf("%d columns", len(ls)),
+					fmt.Sprintf("%d columns", len(rs)),
+				}
+				return false
+			}
+			for i := range ls {
+				if !reflect.DeepEqual(ls[i].Type, rs[i].Type) {
```

Comment:
Seems like this is mutually exclusive with the type conversion code in mergeUnionSchemas

Reply:
It is, but it's supposed to be a sanity check step in case schema merge fails.